### PR TITLE
[espnet3-13] Add logging utils

### PR DESCRIPTION
**Pull request by Masao-Someki**
## What did you change?

Added logging utility and attached loggers.

---

## Why did you make this change?

To enhance the logging feature of espnet3.

---

## Is your PR small enough?

yes, 18 files and -600 line edits.

---

## Additional Context

This PR should be merged after #6328.

* #6133
* #6323
* #6328

### CODE REVIEW

**Comment by Masao-Someki** [comment]
I will add a sample full log with the librispeech-100h recipe in the later comment. It covers from downloading the ls-100h dataset and generating metrics.

**Comment by Masao-Someki** [comment]
With these logging/download utils, we get the logs like this.
It coverts environment variables, git commit id, resolved configs

[log_sample.txt](https://github.com/user-attachments/files/24349180/log_sample.txt)

**Comment by Masao-Someki** [comment]
The logging utility covers a wide range of information, including the execution command, Git commit/branch, cluster environment variables, runtime environment, and the resulved train, inference, and evaluation configs.

[log_sample.txt](https://github.com/user-attachments/files/24349296/log_sample.txt)

**Comment by LiChenda** [comment]
I can review this PR.

**Comment by Masao-Someki** [comment]
I have added the following items to the logging utilities:

1. Added the Python version as metadata.
2. Added a flag to generate `requirements.txt` for experiment-level environment logging.
3. Added a log rotation function for cases where the log file already exists (in espnet2, this was previously handled by a Perl script).

**Note**: We are currently working on separating log files at the stage level (`train.log`, `infer.log`, etc). This will be implemented in PR #6331, since log file specification and generation are handled by `run.py` in each recipe template.

[Log sample from integration test](https://github.com/user-attachments/files/24537697/run.log)

**screenshots**
| number | screenshot |
| --- | --- |
| 1 | image |
| 2 | image |
| 3 | image|

**Comment by sw005320** [comment]
The training log information is very sparse.
It should provide more information

File: espnet3/components/data/dataset.py
Comment (sw005320) [inline espnet3/components/data/dataset.py:1]: Why is this included in this PR?

Comment (Masao-Someki) [reply espnet3/components/data/dataset.py:1]: Thank you, it seems this is a mistake on merging a branch. I will re-check all modified files.

**Comment by Masao-Someki** [comment]
Based on this PR, we would get training log like this: [train.log](https://github.com/user-attachments/files/25159817/train.log)

It generates espnet2-style log during training while plotting progress bar in the shell.

The logs now explicitly include:
* **Run metadata**: command, configs, working directory, start time
* **Code state**: git commit / branch / dirty status
* **Runtime environment**: Python, PyTorch, CUDA/cuDNN, distributed & cluster env
* **Model summary**: parameter counts, dtype, model size, module breakdown
  ```
   Model summary:
      Class Name: ESPnetASRModel
     Total Number of model parameters: 38.47 M
     Number of trainable parameters: 38.47 M (100.0%)
     Size: 153.89 MB
     Type: torch.float32(100.0%)
  ```
* **Optimizer & scheduler**: hyperparameters and effective learning rates
* **DataLoader info**: dataset, sampler, and key loader settings
  ```
  [DataLoader]
  dataset           : espnet3.components.data.dataset.CombinedDataset(len=5567)
  batch_size        : None
  shuffle           : None
  sampler           : torch.utils.data.sampler.SequentialSampler
  batch_sampler     : builtins.list
  num_workers       : 0
  drop_last         : False
  pin_memory        : False
  persistent_workers: False
  prefetch_factor   : None
  timeout           : 0
  multiprocessing   : None
  collate_fn        : espnet2.train.collate_fn.CommonCollateFn
  ```
* **Training progress**: espnet2 style metrics and epoch-level summaries

This makes it much easier to debug past training runs by looking at a single log file.

**Comment by sw005320** [comment]
This looks good to me.
Why is the DataLoader info formatted while the others (e.g., optimizer, metric, inference, etc.) are not?
I recommend that you unify them with non-formatted ones or formatted ones, depending on the difficulty of "the other developers" (not you) to follow

Also, `[main] Infer config content:` --> `[main] inference config content:`

Can you fix the naming issues one by one?

**Comment by Masao-Someki** [comment]
I have prepared some examples of error logs:

- Error in model definition: [model_definition.log](https://github.com/user-attachments/files/25186885/model_definition.log)
- Error in dataset definition: [dataset_define.log](https://github.com/user-attachments/files/25186889/dataset_define.log)
- Error in `__getitem__` in dataset class: [dataset_getitem.log](https://github.com/user-attachments/files/25186897/dataset_getitem.log)
- Invalid configuration (batch_type == numelsort): [invalid_config.log](https://github.com/user-attachments/files/25186911/invalid_config.log)
- Invalid configuration (failure to resolve config values): [invalid_config2.log](https://github.com/user-attachments/files/25186915/invalid_config2.log)
- Invalid `_target_` path: [invalid_target.log](https://github.com/user-attachments/files/25186920/invalid_target.log)

File: espnet3/components/callbacks/default_callbacks.py [hunk 1]
```diff
@@ -122,6 +124,191 @@ def on_validation_end(self, trainer, pl_module):
...
+      - Learning rates (``optim{idx}_lr{group}``)
+
+    Example:
+        >>> cb = TrainBatchMetricsLogger(log_every_n_steps=200)
+        >>> trainer = Trainer(callbacks=[cb, ...])
+
+    Example log output:
+        20epoch:train:4201-4400batch: iter_time=6.212e-05, forward_time=0.145, \
+backward_time=0.159, optim_step_time=0.015, train_time=0.562, \
+loss_ctc=69.86, loss_att=32.128, acc=0.868, loss=46.669, \
```
Comment (LiChenda) [inline espnet3/components/callbacks/default_callbacks.py]: These lines can be indented a little.

File: espnet3/components/data/dataloader.py [hunk 2]
```diff
@@ -173,15 +178,23 @@ def _build_standard_dataloader(self, dataloader_config, dataset=None):
...
+            logger,
+            loader,
+            label=mode,
+            sampler=sampler,
+            batch_sampler=batch_sampler,
+        )
+        return loader
 
-    def _build_iter_factory(self, factory_config, dataset=None):
+    def _build_iter_factory(self, factory_config, dataset=None, *, mode: str):
```
Comment (LiChenda) [inline espnet3/components/data/dataloader.py]: I am not sure if a default value is needed here?

File: espnet3/utils/logging_utils.py [hunk 3]
```diff
@@ -0,0 +1,994 @@
...
+from humanfriendly import format_number, format_size
+
+from espnet3.components.data.dataset import CombinedDataset, DatasetWithTransform
+from espnet3.components.optimizers.multiple_optimizer import MultipleOptimizer
+
+LOG_FORMAT = (
+    "[%(hostname)s] %(asctime)s (%(filename)s:%(lineno)d) "
+    "%(levelname)s:\t[%(stage)s] %(message)s"
+)
+DATE_FORMAT = "%Y-%m-%d %H:%M:%S"
```
Comment (LiChenda) [inline espnet3/utils/logging_utils.py]: Maybe it would be better to make it user-configurable? (Just a suggestion, the current form is good enough)

File: espnet3/utils/logging_utils.py [hunk 4]
```diff
@@ -0,0 +1,994 @@
...
+        log_run_metadata(
+            logger,
+            argv=["espnet3-train", "--config", "conf/train.yaml"],
+        configs={"train": Path("conf/train.yaml")},
+        )
+        ```
+
+    Example log output (wrapped for readability, <= 88 chars):
+        ```
+        2026-02-04 10:15:22 | INFO | espnet3 | === ESPnet3 run started: \
```
Comment (LiChenda) [inline espnet3/utils/logging_utils.py]: Perhaps it could be considered to add the display of time zones?

Comment (sw005320) [reply espnet3/utils/logging_utils.py]: I like the idea!

File: espnet3/components/data/dataloader.py (hunk 2 above)
Comment (Masao-Someki) [reply espnet3/components/data/dataloader.py]: We actually don't need it, but it's confusing as well. I will remove `*,`!

File: espnet3/utils/logging_utils.py (hunk 3 above)
Comment (Masao-Someki) [reply espnet3/utils/logging_utils.py]: I will add a function to override it!

**Comment by Masao-Someki** [comment]
I have updated the branch based on the reviews. Thank you, @sw005320 and @LiChenda!
I'll share the updated log files for `train`, `infer`, `metric` stages.

The changes include the following:
- **Fixed naming issues**
  - `[main] Infer config content:` <- This item comes from the next PR (integration test). I will fix it in that PR.
-  **Unified logging behavior**
  - Removed all custom log formatting and switched to `log_component` based logging. This function searches all attributes of an instance and logs their values.
   - This modification is based on the following comment:
    > Why is the DataLoader info formatted while the others (e.g., optimizer, metric, inference, etc.) are not?
    > I recommend that you unify them with non-formatted ones or formatted ones, depending on the difficulty of "the other developers" (not you) to follow
   - Logging functions are now moved into their respective modules (e.g., DataLoader logging lives in the DataLoader module, `ESPnetLightningModule` logging is handled within the class methods, etc.).
 - Removed `*,` from espnet3 for readability.
 - Other reviews from chenda.

**Comment by Masao-Someki** [comment]
Here's the sample logs with the latest commit from this PR.

- [train.log](https://github.com/user-attachments/files/25215716/train.log)
- [infer.log](https://github.com/user-attachments/files/25215715/infer.log)
- [metric.log](https://github.com/user-attachments/files/25215712/metric.log)

**Comment by sw005320** [comment]
I think you should use `measure` instead of `metric`

File: espnet3/systems/asr/system.py [hunk 5]
```diff
@@ -69,6 +69,66 @@ def create_dataset(self, *args, **kwargs):
         )
         return result
 
+    def _get_stage_log_dirs(self) -> dict[str, Path]:
```
Comment (sw005320) [inline espnet3/systems/asr/system.py]: Do we need to add these lines for each system?

Comment (sw005320) [reply espnet3/systems/asr/system.py]: Can we move some of them to `base/system.py`?

**Comment by Masao-Someki** [comment]
I have updated the names and the `_get_stage_log_dirs` part. Now we can use the default logging path and override it from each systems in `super().__init__()` function.

Here are the updated version of the logs:
- [train.log](https://github.com/user-attachments/files/25230441/train.log)
- [infer.log](https://github.com/user-attachments/files/25230442/infer.log)
- [measure.log](https://github.com/user-attachments/files/25230439/measure.log)

File: espnet3/utils/logging_utils.py [hunk 6]
```diff
@@ -0,0 +1,894 @@
...
+            2026-02-04T10:15:22 ===
+        2026-02-04 10:15:22 | INFO | espnet3 | Command: /usr/bin/python3 \
+            espnet3-train --config conf/train.yaml
+        2026-02-04 10:15:22 | INFO | espnet3 | Python: 3.10.12 (GCC 11.4.0)
+        2026-02-04 10:15:22 | INFO | espnet3 | Working directory: /home/user/espnet3
+        2026-02-04 10:15:22 | INFO | espnet3 | train config: /home/user/espnet3/conf/\
+            train.yaml
+        2026-02-04 10:15:22 | INFO | espnet3 | Git: commit=0123456789abcdef, \
+            short_commit=0123456, branch=main, worktree=clean
```
Comment (sw005320) [inline espnet3/utils/logging_utils.py]: They are very different from the current log output

File: espnet3/utils/logging_utils.py
Comment (sw005320) [inline espnet3/utils/logging_utils.py:1]: Please make sure the comments in the code are up to date.

**Comment by sw005320** [comment]
What is the current status of this PR?

**Comment by Masao-Someki** [comment]
We have fixed the previous review comments, and this PR was pended by CI issues. Now we have fixed the CI issues, once the current CI checks pass, it should be ready to merge.

**Comment by sw005320** [comment]
LGTM.
Once it passes the CI test, you can merge it.
